### ISSUE TITLE
match_like_matches_macro reporting false refactor if one of the arm contains call to macro

### ISSUE THREAD

**Issue by arthurlm** [rust-lang/rust-clippy#7310]
Lint name: `match_like_matches_macro`

I tried this code:

```rust
#![allow(dead_code)]

enum Foobar {
    Foo,
    Bar,
    Baz,
}

const fn foobar() -> Foobar {
    Foobar::Bar
}

fn main() {
    let _ = match foobar() {
        Foobar::Foo => true,
        Foobar::Bar => cfg!(unix),
        _ => false,
    };
}
```

I expected to see this happen:

Nothing. This code looks good to me.

Instead, this happened:

```
warning: match expression looks like `matches!` macro
  --> src/main.rs:14:13
   |
14 |       let _ = match foobar() {
   |  _____________^
15 | |         Foobar::Foo => true,
16 | |         Foobar::Bar => cfg!(unix),
17 | |         _ => false,
18 | |     };
   | |_____^ help: try this: `matches!(foobar(), Foobar::Foo | Foobar::Bar)`
   |
   = note: `#[warn(clippy::match_like_matches_macro)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_like_matches_macro

warning: 1 warning emitted
```

Replacement suggested by clippy `matches!(foobar(), Foobar::Foo | Foobar::Bar)` is only valid on unix platform.
One solution could be to replace the code as bellow:

```rust
let _ = if cfg!(unix) {
	matches!(foobar(), Foobar::Foo | Foobar::Bar)
} else {
	matches!(foobar(), Foobar::Foo)
};
```

But this is less readable and became dangerous when there is a lot of enum items or cfg attributes.
So, to me, clippy should not suggest replacement if one of the arm contains macro call.

### Meta

- `cargo clippy -V`: clippy 0.1.52 (9bc8c42 2021-05-09)
- `rustc -Vv`:
```
rustc 1.52.1 (9bc8c42bb 2021-05-09)
binary: rustc
commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
commit-date: 2021-05-09
host: x86_64-unknown-linux-gnu
release: 1.52.1
LLVM version: 12.0.0
```

**Comment by alex-700** [comment]
The analogue of this problem described at the Known problems section of [lint](https://rust-lang.github.io/rust-clippy/master/index.html#match_like_matches_macro).

This can not be fixed easily, because this lint works for IR where all `cfg` macros was already expanded.

**Comment by arthurlm** [comment]
Thanks for your reply.
I understand this problem may be hard to resolve since cfg macro are expanded before lint is checked.

Indeed, it is mentioned in know problems section of lint (and I did not see it first).

I will add `#[allow(clippy::match_like_matches_macro)]`
I am closing this issue even if it is not really resolved.
